program-runtime/src/invoke_context.rs

```diff
@@ -976,34 +1000,35 @@ impl<'a> InvokeContext<'a> {
     /// Calls the instruction's program entrypoint method
     fn process_executable_chain(&mut self) -> Result<(), InstructionError> {
         let instruction_context = self.transaction_context.get_current_instruction_context()?;
-        let borrowed_root_account = instruction_context
-            .try_borrow_account(self.transaction_context, 0)
-            .map_err(|_| InstructionError::UnsupportedProgramId)?;
-        let root_id = borrowed_root_account.get_key();
-        let owner_id = borrowed_root_account.get_owner();
-        if solana_sdk::native_loader::check_id(owner_id) {
-            for entry in self.builtin_programs {
-                if entry.program_id == *root_id {
-                    drop(borrowed_root_account);
-                    // Call the builtin program
-                    return (entry.process_instruction)(
-                        1, // root_id to be skipped
-                        self,
-                    );
-                }
+
+        let (first_instruction_account, builtin_id) = {
+            let borrowed_root_account = instruction_context
+                .try_borrow_account(self.transaction_context, 0)
+                .map_err(|_| InstructionError::UnsupportedProgramId)?;
+            let root_id = borrowed_root_account.get_key();
+            let owner_id = borrowed_root_account.get_owner();
+            if solana_sdk::native_loader::check_id(owner_id) {
+                (1, *root_id)
+            } else {
+                (0, *owner_id)
             }
-        } else {
-            for entry in self.builtin_programs {
-                if entry.program_id == *owner_id {
-                    drop(borrowed_root_account);
-                    // Call the program via a builtin loader
-                    return (entry.process_instruction)(
-                        0, // no root_id was provided
+        };
+
+        for entry in self.builtin_programs {
+            if entry.program_id == builtin_id {
+                let program_id = instruction_context.get_program_id(self.transaction_context);
+                if builtin_id == program_id {
+                    return process_instruction_with_program_logging(
```

Comment:
Seems like this is the only place where `process_instruction_with_program_logging()` is used now.
Why not inline it? Then most of the redundant code to get the `program_id` could be removed as well.

Reply:
Yeah great call, it's much cleaner inlined